pkg/kubelet/status/state/checkpoint.go

```diff
@@ -18,48 +18,62 @@ package state
 
 import (
 	"encoding/json"
-
 	"k8s.io/api/core/v1"
 	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
 	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager/checksum"
 )
 
-var _ checkpointmanager.Checkpoint = &PodResourceAllocationCheckpoint{}
+var _ checkpointmanager.Checkpoint = &Checkpoint{}
 
-// PodResourceAllocationCheckpoint is used to store resources allocated to a pod in checkpoint
-type PodResourceAllocationCheckpoint struct {
+type PodResourceAllocationInfo struct {
 	AllocationEntries   map[string]map[string]v1.ResourceList `json:"allocationEntries,omitempty"`
 	ResizeStatusEntries map[string]v1.PodResizeStatus         `json:"resizeStatusEntries,omitempty"`
-	Checksum            checksum.Checksum                     `json:"checksum"`
 }
 
-// NewPodResourceAllocationCheckpoint returns an instance of Checkpoint
-func NewPodResourceAllocationCheckpoint() *PodResourceAllocationCheckpoint {
-	//lint:ignore unexported-type-in-api user-facing error message
-	return &PodResourceAllocationCheckpoint{
-		AllocationEntries:   make(map[string]map[string]v1.ResourceList),
-		ResizeStatusEntries: make(map[string]v1.PodResizeStatus),
+// Checkpoint represents a structure to store pod resource allocation checkpoint data
+type Checkpoint struct {
+	// Data is a JSON serialized checkpoint data
+	Data string `json:"data"`
+	// Checksum is a checksum of Data
+	Checksum checksum.Checksum `json:"checksum"`
+}
```

Comment:
Changing checkpoint format should not cause Kubelet failures. When older Kubelet creates checkpoint and then upgraded to the new version with this format, it will not start for the same reason as you're trying to fix with this PR.

Reply:
It would be nice to test upgrade scenario, e.g. if Data has a new field(it could happen when Kubelet is upgraded), but the checkpoint is not(as it's created with older Kubelet, before the upgrade).

Reply:
Do you mean we should have a explicit test case to expose that state checkpoint will fail when Data has a new field, like Kubelet is upgraded?

Reply:
The point is that Kubelet shouldn't fail. The test case should demonstrate that Kubelet is able to read checkpoints created before Kubelet upgrades.

Reply:
@yunwang0911 here is an example of such a test case for DRA checkpointing: https://github.com/kubernetes/kubernetes/pull/126717

Reply:
This should be addressed. The rest of the PR looks good enough to me.

Reply:
@bart0sh I added another test case to test upgrade scenario. https://github.com/kubernetes/kubernetes/pull/126620/files#diff-ff3988755adc35924ce7f3bf6ffa770cd159b34af6428db5f0f9c461fd15234fR135
If I understand it wrong, please kindly correct me

Reply:
@SergeyKanzhelev In theory, this issue only affects static pods. Because the marshal and un-marshal will be invoked when the data is serialized and deserialized. Please correct me if I'm wrong.